### DISCUSSION TITLE
How to check for `undefined`

### DISCUSSION THREAD

**Discussion by mxdvl** [guardian/csnx#442] [Team Posts]
There are currently three ways of checking whether a value is `undefined` in JavaScript:

```ts
// strict equality
if(variable === void 0) // shortest, but we seldom use it
if(variable === undefined) // we use this a lot

// type check
if(typeof variable === "undefined") // safest
 
```

We use the last two in our codebases, but thankfully since [ES5] the global `undefined` cannot be redefined. It can still be shadowed, but the recommended ESLint rules prevent this via [`no-shadow-restricted-names`][eslint].

Due to this, I believe **it is not necessary to enforce this rule** globally.

[ES5]: https://262.ecma-international.org/5.1/#sec-15.1.1.3
[eslint]: https://eslint.org/docs/rules/no-shadow-restricted-names

### Appendix

To get TypeScript to play nice, you need to use [type predicates][tp] and pass it as the filter. [And here’s a working example][ex].

[tp]: https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates
[ex]: https://www.typescriptlang.org/play?#code/MYewdgzgLgBAtgTwPICMBWBTYsC8MDeAsAFACQ4GAXDAIwA0JpUA7iNQOQsjsNkBmIAK4AnaoLAATDHwCWYDBN6lZANyow+AQwA2EDEogyAHtXwwK1LbowwAvr1sBuEiVCRYs7VAzCFqTNgwOIz+WFAAdBhgUMIyGBAAFIih2ACUjOGe3sIJCQDadDAqOoIYALqpQQB8RSU2OA0w4lKy8hKpzsSu4NAwABaaEADKMXIA5kjCAEIgINoYmmAAanUQQTAJVHnQsWBjheIA1mAgzGAV1DYya9uje4U74zAAPjBggnAoPmXVjHkARI89v9Cv93p8fP8yuE5MBtIIpIkoAgAA4YEB8GAYPI0CqdEgAegJMBRwhAaOE2gQADEZF4fH50GFqLddvsYECxi83h8vsIynkyt13CSyRSqbT6b4JClcCEmdhItFYvEksgFVB0mRMnTsgkBsM7hNprN5osVvD4lrhb1wFSAKqSaRyBQW0prPCbSis8YHMDHU7nVKXGDXGA++5NJ2tBQ-HBVRjY3FBRrNZ1tfHEIlRloumUalmcv25toCoXENy9NMx-MBXAwWVK0aq5IarWkHVShJ2hCOkuu1ZaoA

**Comment by sndrs** [comment]
great digging, thanks!

**Comment by sndrs** [comment]
for future travellers https://github.com/guardian/libs/pull/211
